Enable user defined sample and container geometry for abs correction

Kvieta1990 opened the pull request:
### Description of work

#### Summary of work
With this PR, we enable user defined sample and container geometry for absorption correction. Top level changes are in place for `SNSPowderReduction` and accordingly necessary parameters are introduced for those bottom level algorithms concerning the setup of the absorption correction inputs. The defined sample and container geometry is finally passed to `SetSample`. This gives users the flexibility to define their own geometry as needed. For example, on POWGEN, some times irregular containers that are not defined in the sample environment XML file will be used for the measurement, e.g., the annular container where the sample is loaded into the hollow cylinder wall.

Also, we implement the definition of beam gauge volume for absorption correction. Without this implementation, the assumption would be that the beam size is the same as the sample size. In practice, this is often not true. In that case, the absorption correction is imperfect in principle. Detailed discussion about this is posted [here](https://iris2020.net/2025-02-12-abs_geo/).

### To test:

#### This is the one for testing the gauge volume definition

```python
from mantid.simpleapi import *

gauge_vol = '''<cuboid id="shape">
     <left-front-bottom-point x="0.01" y="-0.1" z="0.0"  />
     <left-front-top-point  x="0.01" y="-0.1" z="0.02"  />
     <left-back-bottom-point  x="-0.01" y="-0.1" z="0.0"  />
     <right-front-bottom-point  x="0.01" y="0.1" z="0.0"  />
   </cuboid>'''

SNSPowderReduction(
    Filename='/SNS/PG3/IPTS-34523/nexus/PG3_58812.nxs.h5',
    CalibrationFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_OC_d58772_2025-02-12.h5',
    CharacterizationRunsFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2-HighRes_OC.txt,/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2_OC_limit.txt',
    RemovePromptPulseWidth=50,
    Binning='-0.0008',
    BackgroundSmoothParams='5, 2',
    FilterBadPulses=90,
    ScaleData=100,
    TypeOfCorrection='SampleAndContainer',
    SampleFormula='Ba-Nb-Cr-Ru',
    MeasuredMassDensity='6.0',
    SampleGeometry={'Height': 5},
    SampleNumberDensity='0.088',
    ContainerShape='PAC06',
    GaugeVolume=gauge_vol,
    SaveAs='gsas topas and fullprof',
    OutputDirectory='/SNS/NOM/shared/tmp',
    CacheDir='/SNS/NOM/shared/tmp'
)
```

#### Here is the one for testing the geometry definition,

```python
from mantid.simpleapi import *

sam_geo = {
    'Shape': 'HollowCylinder', 'Height': 4.0,
    'InnerRadius': 2.0,
    'OuterRadius': 3.0,
    'Center': [0.,0.,0.]
}

c_geo = {
    'Shape': 'HollowCylinderHolder',
    'Height': 4.0,
    'InnerRadius': 1.5,
    'InnerOuterRadius': 2.0,
    'OuterInnerRadius': 3.0,
    'OuterRadius': 4.0,
    'Center': [0.,0.,0.]
}

c_mat = {
    'ChemicalFormula': 'V',
    'NumberDensity': 0.072324
}

gauge_vol = '''<cuboid id="shape">
     <left-front-bottom-point x="0.01" y="-0.1" z="0.0"  />
     <left-front-top-point  x="0.01" y="-0.1" z="0.02"  />
     <left-back-bottom-point  x="-0.01" y="-0.1" z="0.0"  />
     <right-front-bottom-point  x="0.01" y="0.1" z="0.0"  />
   </cuboid>'''

SNSPowderReduction(
    Filename='/SNS/PG3/IPTS-34523/nexus/PG3_58812.nxs.h5',
    CalibrationFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_OC_d58772_2025-02-12.h5',
    CharacterizationRunsFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2-HighRes_OC.txt,/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2_OC_limit.txt',
    RemovePromptPulseWidth=50,
    Binning='-0.0008',
    BackgroundSmoothParams='5, 2',
    FilterBadPulses=90,
    ScaleData=100,
    TypeOfCorrection='SampleAndContainer',
    SampleFormula='Ba-Nb-Cr-Ru',
    MeasuredMassDensity='6.0',
    SampleGeometry=sam_geo,
    SampleNumberDensity='0.088',
    ContainerGeometry=c_geo,
    ContainerMaterial=c_mat,
    GaugeVolume=gauge_vol,
    SaveAs='gsas topas and fullprof',
    OutputDirectory='/SNS/NOM/shared/tmp',
    CacheDir='/SNS/NOM/shared/tmp'
)
```

#### Here is another one for testing the geometry definition, with only beam height defined (and therefore the gauge volume will be defined internally in Mantid by assuming the beam location),

```python
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

SNSPowderReduction(
    Filename='/SNS/PG3/IPTS-34523/nexus/PG3_58812.nxs.h5',
    CalibrationFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_OC_d58772_2025-02-12.h5',
    CharacterizationRunsFile='/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2-HighRes_OC.txt,/SNS/PG3/shared/CALIBRATION/2025-1_11A_CAL/PG3_char_2025_2_OC_limit.txt',
    RemovePromptPulseWidth=50,
    Binning='-0.0008',
    BackgroundSmoothParams='5, 2',
    FilterBadPulses=90,
    ScaleData=100,
    TypeOfCorrection='SampleAndContainer',
    SampleFormula='Ba-Nb-Cr-Ru',
    MeasuredMassDensity='6.0',
    SampleGeometry={
        'Height': 5,
        'Radius': 0.295
    },
    BeamHeight=2,
    SampleNumberDensity='0.088',
    ContainerShape='PAC06',
    ContainerGaugeVolume='0.295 0.315',
    SaveAs='gsas topas and fullprof',
    OutputDirectory='/SNS/NOM/shared/tmp',
    CacheDir='/SNS/NOM/shared/tmp'
)
```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Kvieta1990 committed:
enable user defined sample and container geometry for abs correction

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Kvieta1990 committed:
add in release doc

Kvieta1990 committed:
Merge branch 'abs_can_geo_def' of https://github.com/mantidproject/mantid into abs_can_geo_def

Kvieta1990 committed:
fix logic

Kvieta1990 committed:
fix beam height issue

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
remove redundant checks

Commit:
fix gauge volume issue

Kvieta1990 commented on Framework/PythonInterface/mantid/utils/absorptioncorrutils.py:
```diff
@@ -607,11 +638,31 @@ def confirmProps(props):
         )
 
+    if beam_height != Property.EMPTY_DBL and not gauge_vol:
+        gauge_vol = """<cuboid id="shape">
```
The [DefineGaugeVolume user docs](https://docs.mantidproject.org/nightly/algorithms/DefineGaugeVolume-v1.html) is indeed a bit confusing in the explanation of how the gauge volume is defined. Here are several comments,

1. For the cuboid volume definition, I created an image to demonstrate the idea,
![Cuboid volume definition demo](https://iris2020.net/assets/img/posts/Mantid_Gauge_Volume_Dev.png "Cuboid volume definition demo")

2. Ideally, indeed the gauge volume could be defined as above a a cuboid and Mantid then worries about constructing the intersection area between the sample and the defined gauge volume. However, the construction of the intersection area is not implemented yet. So, we cannot go with this route.

3. However, the gauge volume can be defined as with the same shape as the sample, e.g., cylinder or hollow cylinder and current implementation can meet our need without problems. We do need some further implementation in this PR to define the gauge volume for the sample and container separately.

4. The implementation for `SampleOnly` and `SampleAndContainer` method is straightforward -- just define the gauge volume twice [here](https://github.com/mantidproject/mantid/blob/de114fd17e7a23cf0f3ca7a2ec6fd9064eba7cba/Framework/PythonInterface/mantid/utils/absorptioncorrutils.py#L484-L485)

5. For the `PaalmanPingsAbsorptionCorrection` method, I am not sure how to do that  -- the same principle applies but I am not so familiar with the CPP codes.

RichardWaiteSTFC replied:
At ISIS we are also interested in starting to use `DefineGaugeVolume` with other attenuation algorithms such as `MonteCarloAbsoption` for use on our engineering beamline ENGIN-X. For us we would need to consider the case of when the gauge volume is not wholly occupied by the sample, both for attenuation correction and normalising by the illuminated sample volume. If you do implement this let us know! Not sure how this fits in with your plans? Happy to discuss!

Kvieta1990 replied:
sure. I will keep you updated about the implementation. We do plan to do the gauge volume definition for both the absorption correction and normalization purpose. Currently I am working on this and the PR here is part of the work. Maybe we need another PR for working with implementation for `PaalmanPingsAbsorptionCorrection` which may need the help from @peterfpeterson and his team since I am not so familiar the C++ code and the code architecture there^_^ Anyhow, I will keep you in the loop.

peterfpeterson replied:
There is also the algorithm `SetBeam` which may be appropriate for these calculations. I'm still in progress on understanding the effect of it.

Commit:
container gauge vol definition

Kvieta1990 commented:
@peterfpeterson I put in the implementation to define the gauge volume for sample and container separately to account for the beam size. I think you are right -- the way that gauge volume definition is that the defined gauge volume will just replace the part of the sample shined by the beam. I did the test with a gauge volume larger then the sample size and in that case, as we change the gauge volume, the absorption spectrum is changing accordingly. This indicates that the gauge volume is not working as what I thought it would work, i.e., calculate the intersection region between the sample and the defined gauge volume.

As I put above, I think this is fine. We then just need to define the gauge volume to follow exactly the shape of the sample but with different height and position so only the defined gauge volume region will be taken as the integration volume. When calculating L2 (i.e., the path after the scattering all the way to detectors), still the original sample volume will be used -- I think this is how the codes are currently working but it will be nice to double check to make sure. We also need to do the same thing for the container separately, just like what we did with the sample, and the gauge volume for sure needs to be defined separately for the container to follow its shape (usually, a hollow cylinder).

All these can be done in the absorption util script for `SampleOnly` and `SampleAndContainer` types of correction. For `PaalmanPingsAbsorptionCorrection`, it is buried in the C++ implementation and I kind of lose myself in the code base and I may need your help on this. I will write a story for this. Tried to do that last night but the IBM tool seems to be problematic blocking me from logging in. Will do that later so we can schedule it next week or some time recently. This is a bit urgent as we do rely on a proper absorption correction for reducing samples with strong absorption quite often on both NOMAD and POWGEN. Thank you!

Kvieta1990 commented:
@peterfpeterson could you please leave this open while I am testing to see whether the currently implementation is doing its job properly? Or should I just change to draft?

peterfpeterson reviewed (approved):
This appears to be right for simple absorption calculations. It is unclear if this would work for Paalman-Pings with the current state of the calculations. 

tl;dr use for simple absorption only

peterfpeterson reviewed (changes requested):
Blocking for now for more testing

Commit:
fix quite a few issues with the implementation

Commit:
fix vanadium gauge volume def issue

Kvieta1990 commented:
> Blocking for now for more testing

Hi @peterfpeterson I just tested this out, making some more changes to the code and from the test result, it seems to be working fine. For the new changes,

- I put in some unit conversion internally so that users will always put in length in cm.

- I fix the issue with the previous XML string format -- `{:4.2F}` was setting the gauge volume radius to `0` as the real value should be something like `0.00295` (with XML, it requires length in `m`).

- For the container part of the absorption calculation, the previous implementation had some issue -- if the container gauge volume is not defined, it will use the gauge volume defined for the sample since they share the same donor workspace and the gauge volume is defined on the donor workspace only. So, if the sample gauge volume is defined, for the `SampleAndContainer` type of correction, we have to define the gauge volume for the container as well. I added this in with the latest commits.

- I also noticed that the previous implementation missed the gauge volume definition for vanadium. I also added that in.

RichardWaiteSTFC commented on Framework/PythonInterface/mantid/utils/absorptioncorrutils.py:
```diff
@@ -611,7 +679,31 @@ def confirmProps(props):
+        # and we will be assuming a cylinder shape of the sample.
+        gauge_vol = """<cylinder id="shape">
+            <centre-of-bottom-base r="{0:4.2F}" t="90.0" p="270.0" />
+            <axis x="0.0" y="0.2" z="0.0" />
```
I believe it's not always the case that the vertical is along the y-axis - I think you can get this from the IDF/instrument?

Kvieta1990 replied:
Sorry about the delay on my side due to local contacting experiments. I am not sure what is the best way to obtain the axis direction. Could you please help provide some hints about how specifically we can do the checking? Thanks!

RichardWaiteSTFC replied:
No worries, you can access this from the instrument reference frame like so
https://github.com/mantidproject/mantid/blob/3306871804d5a749a7c0fb248d18f7eab83812ca/Testing/SystemTests/tests/framework/RelectometryInstrumentSignedThetaTest.py#L25-L30

RichardWaiteSTFC replied:
But if you know all the instruments using this script have the y-axis vertical then I'm happy to keep as it is!

Kvieta1990 replied:
> But if you know all the instruments using this script have the y-axis vertical then I'm happy to keep as it is!

To be honest, I am not sure at this moment. I will check with @peterfpeterson and folks here to see. But if the implementation for incorporating the instrument definition is not so involved, I will do it anyhow. Will see and will let you know. Thanks a ton!

RichardWaiteSTFC reviewed (changes requested):
`absorptioncorrutils` is not used in the ISIS powder reduction -  so I don't want to hold this up! Here are a few suggestions/questions. Also it seems like some of the PR testing instructions could be used to make a system test? However if this work is in-progress/beta I understand wanting to wait a bit until it's been validated.

Kvieta1990 commented on Framework/PythonInterface/mantid/utils/absorptioncorrutils.py:
```diff
@@ -611,7 +679,31 @@ def confirmProps(props):
+            sam_rad = geometry["Radius"]
+        else:
+            sam_rad = geometry["Radius"].value
+        gauge_vol = gauge_vol.format(beam_height / 200.0, sam_rad / 100.0, beam_height / 100.0)
```
When defining the cylinder geometry, I followed the instruction on Mantid documentation page which says `centre-of-bottom-base` is using the spherical coordinate system. `t=90` means the xy plane, `p=270` means the negative y direction and the division factor `2` for `r` means I want to put the center of the bottom base to half of the cylinder height along the negative y direction so the center of the gauge volume is at the coordinate origin. This is the an assumption if only the beam size is provided.

Yes, `100` factor there is for unit conversion from `cm` to `m`.

Kvieta1990 replied:
sure thing!

Commit:
add in gauge volume exception

Kvieta1990 commented on Framework/PythonInterface/mantid/utils/absorptioncorrutils.py:
```diff
@@ -473,6 +507,13 @@ def calc_1st_absorption_corr_using_wksp(
             raise RuntimeError("Specified donor workspace not found in the ADS")
         donor_wksp = mtd[donor_wksp]
 
+    def is_valid_xml(xml_string):
```
I put in a `try/except` so that if a user defined gauge volume is given explicitly, we will try to use it. If it fails, just skip the gauge volume definition. If the beam height is given, we will not do the `try/except` since the gauge volume is defined internally.

Kvieta1990 commented:
> `absorptioncorrutils` is not used in the ISIS powder reduction - so I don't want to hold this up! Here are a few suggestions/questions. Also it seems like some of the PR testing instructions could be used to make a system test? However if this work is in-progress/beta I understand wanting to wait a bit until it's been validated.

I would push this part of work to another user story that the team here is working on. That work is about a once-for-all solution where the beam itself can be defined and can be used for the gauge volume definition. I think we will create some systematic system tests there. Thanks!

andy-bridger commented:
I've been having a bit of a poke around in the `AbsorptionCorrection` and `DefineGaugeVolume` code because I was hoping to use them for an Engineering Diffraction Texture analysis workflow. I wrote up what I found 'wrong' with the code on this issue: https://github.com/mantidproject/mantid/issues/39079. The caveat, as I mention on there but is also discussed above here, is that if you push explicitly defining the full volume of illuminated sample onto the user, it probably works fine. 

Unfortunately, as far as I am aware, the concept of gauge volume understood within texture analysis is: "the region of the object that is gauged by the instrument corresponds to a cuboid, which results from the intersection of the incident and diffracted beams" [https://doi.org/10.1107/S1600576714012710], so requiring definition of the full illuminated region would not be very intuitive. Additionally if the sample shape is complex, and imported as an STL/mesh, it might be challenging for a user to explicitly define the resultant illuminated volume.

An option I proposed to fix this is to pass both the sample shape and the gauge volume to the Rasterize function and allow the intersection of the two volumes to be handled within. I think this would fix my problem case, but a lower level change to `DefineGaugeVolume`, which has a flag for which region is being passed and calculates full illuminated volumes internally (basically option 2 here: https://github.com/mantidproject/mantid/pull/38887#discussion_r1964846618) might be a preferable solution?

andy-bridger commented:
I guess a concern is, from the `AbsorptionCorrection` view point, using a gauge volume defined by the full volume of illuminated sample will slow the algorithm down quite significantly. You will end up with a significantly larger bounding box for the illuminated area, which will have you rasterising this with N^3 elements, only to throw away a larger chunk of the elements not within the integration volume and a chunk more with invalid paths.

andy-bridger commented:
I don't know if there is an ORNL preference for how the different interpretations of the gauge volume should be handled? @peterfpeterson @Kvieta1990

Kvieta1990 commented:
> I don't know if there is an ORNL preference for how the different interpretations of the gauge volume should be handled? @peterfpeterson @Kvieta1990

We have a user story here to be worked on where we are going to implement the beam definition and using it for defining the gauge volume. With the current PR, we are assuming the user knows what they are doing with the gauge volume definition and it is up to the users to give a proper definition of the gauge volume, i.e., the portion of the sample (and/or container) that is exposed to the beam. Technically, this is totally feasible and I have done tests with real data and it is working fine.

Kvieta1990 commented:
@RichardWaiteSTFC I will finish the new implementation today to account for the coordinate system direction and it will be very nice if you and @peterfpeterson can help approve the PR so that we can use it (for `SampleOnly` and `SampleAndContainer` type of absorption correction). Thanks!

RichardWaiteSTFC commented:
> @RichardWaiteSTFC I will finish the new implementation today to account for the coordinate system direction and it will be very nice if you and @peterfpeterson can help approve the PR so that we can use it (for `SampleOnly` and `SampleAndContainer` type of absorption correction). Thanks!

Yep will do - thanks for the changes.
I think @andy-bridger comments relate to future planned changes we need to get the texture analysis workflow working at ISIS, rather than requests to change this PR. Perhaps it would be possible to arrange a meeting with yourself and @peterfpeterson to discuss?

Kvieta1990 commented:
> > @RichardWaiteSTFC I will finish the new implementation today to account for the coordinate system direction and it will be very nice if you and @peterfpeterson can help approve the PR so that we can use it (for `SampleOnly` and `SampleAndContainer` type of absorption correction). Thanks!
> 
> Yep will do - thanks for the changes. I think @andy-bridger comments relate to future planned changes we need to get the texture analysis workflow working at ISIS, rather than requests to change this PR. Perhaps it would be possible to arrange a meeting with yourself and @peterfpeterson to discuss?

That makes a lot of sense. I will talk to Pete and will get back to you about the meeting for discussion.

Commit:
implemented coord system checking

RichardWaiteSTFC reviewed (approved):
Thanks for the changes, looks good - sorry for delaying this when you're obviously v busy with local contacting! FYI in future I think you can also use f-strings rather than `.format` which is perhaps more readable - but super minor and not worth holding this up!
I'm approving, but I don't have access to the data so can't do functional testing - I'll leave that to ORNL as this change doesn't affect ISIS!

Kvieta1990 commented:
> Thanks for the changes, looks good - sorry for delaying this when you're obviously v busy with local contacting! FYI in future I think you can also use f-strings rather than `.format` which is perhaps more readable - but super minor and not worth holding this up! I'm approving, but I don't have access to the data so can't do functional testing - I'll leave that to ORNL as this change doesn't affect ISIS!

Thanks a lot for all your comments and sure I will be using the `f-strings`! Thanks!
